src/Tools/PrepareTests/TestDiscovery.cs

```diff
@@ -23,88 +17,78 @@ public static void RunDiscovery(string repoRootDirectory, string dotnetPath, boo
         {
-            var directory = Path.GetDirectoryName(assemblyGroup.Key);
-
-            // Tests with combinatorial data are output multiple times with the same fully qualified test name.
-            // We only need to include it once as run all combinations under the same filter.
-            var testToWrite = assemblyGroup.Select(test => test.FullyQualifiedName).Distinct().ToList();
+            var workerPath = assembly.Contains("net472")
+                ? dotnetFrameworkWorker
+                : dotnetCoreWorker;
 
-            using var fileStream = File.Create(Path.Combine(directory!, "testlist.json"));
-            JsonSerializer.Serialize(fileStream, testToWrite);
-        }
+            RunWorker(dotnetPath, workerPath, assembly);
+        });
         stopwatch.Stop();
-        Console.WriteLine($"Serialized tests in {stopwatch.Elapsed}");
+
+        Console.WriteLine($"Discovered tests in {stopwatch.Elapsed}");
     }
 
-    private class DiscoveryHandler : ITestDiscoveryEventsHandler
+    static (string dotnetCoreWorker, string dotnetFrameworkWorker) GetWorkers(string binDirectory)
     {
-        private readonly ConcurrentBag<TestCase> _tests = new();
-        private bool _isComplete = false;
+        var testDiscoveryWorkerFolder = Path.Combine(binDirectory, "TestDiscoveryWorker");
+        var configuration = Directory.Exists(Path.Combine(testDiscoveryWorkerFolder, "Debug")) ? "Debug" : "Release";
+        return (Path.Combine(testDiscoveryWorkerFolder, configuration, "net6.0", "TestDiscoveryWorker.dll"),
+                Path.Combine(testDiscoveryWorkerFolder, configuration, "net472", "TestDiscoveryWorker.exe"));
+    }
 
-        public void HandleDiscoveredTests(IEnumerable<TestCase>? discoveredTestCases)
+    static void RunWorker(string dotnetPath, string pathToWorker, string pathToAssembly)
+    {
+        var pipeClient = new Process();
+        var arguments = new List<string>();
+        if (pathToWorker.EndsWith("dll"))
         {
-            if (discoveredTestCases != null)
-            {
-                foreach (var test in discoveredTestCases)
-                {
-                    _tests.Add(test);
-                }
-            }
+            arguments.Add(pathToWorker);
+            pipeClient.StartInfo.FileName = dotnetPath;
         }
-
-        public void HandleDiscoveryComplete(long totalTests, IEnumerable<TestCase>? lastChunk, bool isAborted)
+        else
         {
-            if (lastChunk != null)
-            {
-                foreach (var test in lastChunk)
-                {
-                    _tests.Add(test);
-                }
-            }
-
-            _isComplete = true;
+            pipeClient.StartInfo.FileName = pathToWorker;
         }
 
-        public void HandleLogMessage(TestMessageLevel level, string? message)
+        using (var pipeServer = new AnonymousPipeServerStream(PipeDirection.Out, HandleInheritability.Inheritable))
```

Comment:
Curious: why did you choose a multi-process model here vs. a multi-thread model?

Reply:
@jmarolf haven't seen an answer to this question